google/cloud/bigquery/opentelemetry_tracing.py

```diff
@@ -0,0 +1,122 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from contextlib import contextmanager
+from google.api_core.exceptions import GoogleAPICallError
+
+Logger = logging.getLogger(__name__)
+try:
+    from opentelemetry import trace
+    from opentelemetry.instrumentation.utils import http_status_to_canonical_code
+    from opentelemetry.trace.status import Status
+
+    HAS_OPENTELEMETRY = True
+
+except ImportError:
+    Logger.info(
+        "This service is instrumented using OpenTelemetry."
+        "OpenTelemetry could not be imported; please"
+        "add opentelemetry-api and opentelemetry-instrumentation"
+        "packages in order to get BigQuery Tracing data."
+    )
+
+    HAS_OPENTELEMETRY = False
+
+_default_attributes = {
+    "db.system": "BigQuery"
+}  # static, default values assigned to all spans
+
+
+@contextmanager
+def create_span(name, attributes=None, client=None, job_ref=None):
+    """Creates a ContextManager for a Span to be exported to the configured exporter.
+        If no configuration exists yields None.
+
+            Args:
+                name (str): Name that will be set for the span being created
+                attributes (Optional[dict]):
+                    Additional attributes that pertain to
+                    the specific API call (i.e. not a default attribute)
+                client (Optional[google.cloud.bigquery.client.Client]):
+                    Pass in a Client object to extract any attributes that may be
+                    relevant to it and add them to the created spans.
+                job_ref (Optional[google.cloud.bigquery.job._AsyncJob])
+                    Pass in a _AsyncJob object to extract any attributes that may be
+                    relevant to it and add them to the created spans.
+
+            Yields:
+                opentelemetry.trace.Span: Yields the newly created Span.
+
+            Raises:
+                google.api_core.exceptions.GoogleAPICallError:
+                    Raised if a span could not be yielded or issue with call to
+                    OpenTelemetry.
+            """
+    final_attributes = _get_final_span_attributes(attributes, client, job_ref)
+    if not HAS_OPENTELEMETRY:
+        yield None
+        return
+    tracer = trace.get_tracer(__name__)
+
+    # yield new span value
+    with tracer.start_as_current_span(name=name, attributes=final_attributes) as span:
+        try:
+            yield span
```

Comment:
I'm a bit confused as to how this line could throw an exception. Could you add a unit test on client or job to verify the span status in this case?

Reply:
Added a test that throws the exception in `test_opentelemetry_tracing.py`. Unless you were referring to a different type of test. That test raises an exception within the context of the span.